File: LeetCode题目集/20. Valid Parentheses(easy).java
/**
 * @Date    : 2019-01-07 20:35:41
 * @Author  : 酸饺子 (changzheng300@foxmail.com)
 * @Link    : https://github.com/SourDumplings
 * @Version : $Id$
 *
 * https://leetcode.com/problems/valid-parentheses/
*/

class Solution
{
    public boolean isValid(String s)
    {
        Stack<Character> S = new Stack<Character>();
        for (int i = 0; i != s.length(); ++i)
        {
            Character c = new Character(s.charAt(i));
            if (c == '(' || c == '[' || c == '{')
            {
                S.push(c);
            }
            else if (c == ')')
            {
                if (!S.empty() && S.peek() == '(')
                {
                    S.pop();
                }
                else
                    return false;
            }
            else if (c == ']')
            {
                if (!S.empty() && S.peek() == '[')
                {
                    S.pop();
                }
                else
                    return false;
            }
            else if (c == '}')
            {
                if (!S.empty() && S.peek() == '{')
                {
                    S.pop();
                }
                else
                    return false;
            }
        }
        if (S.empty())
        {
            return true;
        }
        else
            return false;
    }
}

File: OJ practices/LeetCode题目集/20. Valid Parentheses(easy).cpp
/**
 * @file 20. Valid Parentheses(easy).cpp
 * @author SourDumplings (sourdumplings@qq.com)
 * @brief https://leetcode.cn/problems/valid-parentheses/?envType=study-plan-v2&envId=top-interview-150
 * @version 1.0.0
 * @date 2024-11-13
 * 
 * @copyright Copyright (c) 2024 SourDumplings
 * 
 */

class Solution
{
public:
	bool isValid(string s)
	{
		stack<char> S;
		for (char c : s)
		{
			switch (c)
			{
				case '(':
				case '[':
				case '{':
				{
					S.push(c);
					break;
				}
				case ')':
				{
					if (S.empty() || S.top() != '(') { return false; }
					S.pop();
					break;
				}
				case ']':
				{
					if (S.empty() || S.top() != '[') { return false; }
					S.pop();
					break;
				}
				case '}':
				{
					if (S.empty() || S.top() != '{') { return false; }
					S.pop();
					break;
				}

				default: break;
			}
		}
		return S.empty();
	}
};